File: WordCount.cs
// WordCount.cs - Word frequency counter
// Build: dotnet build -c Release
// Usage: dotnet run --configuration Release [filename]

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;

class WordCount
{
    static void Main(string[] args)
    {
        string filename = args.Length > 0 ? args[0] : "book.txt";
        
        if (!File.Exists(filename))
        {
            Console.Error.WriteLine($"Error: File '{filename}' not found");
            Console.WriteLine("Usage: WordCount.exe [filename]");
            Console.WriteLine("\nTo create a test file:");
            Console.WriteLine("curl https://www.gutenberg.org/files/2701/2701-0.txt -o book.txt");
            Environment.Exit(1);
        }
        
        Console.WriteLine($"Processing file: {filename}");
        
        var stopwatch = Stopwatch.StartNew();
        long startMemory = GC.GetTotalMemory(false);
        
        try
        {
            var (counts, totalWords) = ProcessFileOptimized(filename);
            
            var sorted = counts
                .OrderByDescending(kvp => kvp.Value)
                .ThenBy(kvp => kvp.Key)
                .Take(100)
                .ToList();
            
            stopwatch.Stop();
            long endMemory = GC.GetTotalMemory(false);
            double executionTime = stopwatch.Elapsed.TotalMilliseconds;
            double memoryUsed = (endMemory - startMemory) / 1024.0 / 1024.0;
            double fileSize = new FileInfo(filename).Length / 1024.0 / 1024.0;
            
            Console.WriteLine("\n=== Top 10 Most Frequent Words ===");
            for (int i = 0; i < Math.Min(10, sorted.Count); i++)
            {
                var kvp = sorted[i];
                Console.WriteLine($"{i + 1,2}. {kvp.Key,-15} {kvp.Value,8:N0}");
            }
            
            Console.WriteLine("\n=== Statistics ===");
            Console.WriteLine($"File size:       {fileSize:F2} MB");
            Console.WriteLine($"Total words:     {totalWords:N0}");
            Console.WriteLine($"Unique words:    {counts.Count:N0}");
            Console.WriteLine($"Execution time:  {executionTime:F2} ms");
            Console.WriteLine($"Memory used:     {memoryUsed:F2} MB");
            Console.WriteLine($".NET version:    {Environment.Version}");
            Console.WriteLine($"64-bit process:  {Environment.Is64BitProcess}");
            
            WriteOutputFile(filename, sorted, totalWords, counts.Count, executionTime);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error processing file: {ex.Message}");
            Environment.Exit(1);
        }
    }
    
    static (Dictionary<string, int> counts, int totalWords) ProcessFileOptimized(string filename)
    {
        var counts = new Dictionary<string, int>(20000);
        int totalWords = 0;
        
        ReadOnlySpan<byte> fileBytes = File.ReadAllBytes(filename);
        
        if (fileBytes.Length >= 3 && 
            fileBytes[0] == 0xEF && 
            fileBytes[1] == 0xBB && 
            fileBytes[2] == 0xBF)
        {
            fileBytes = fileBytes.Slice(3);
        }
        
        Span<byte> wordBuffer = stackalloc byte[100];
        int wordLength = 0;
        
        for (int i = 0; i < fileBytes.Length; i++)
        {
            byte b = fileBytes[i];
            
            if ((b >= 65 && b <= 90) || (b >= 97 && b <= 122))
            {
                if (wordLength < wordBuffer.Length)
                {
                    wordBuffer[wordLength++] = (byte)(b >= 65 && b <= 90 ? b + 32 : b);
                }
            }
            else if (wordLength > 0)
            {
                string word = Encoding.ASCII.GetString(wordBuffer.Slice(0, wordLength));
                
                if (counts.TryGetValue(word, out int count))
                {
                    counts[word] = count + 1;
                }
                else
                {
                    counts[word] = 1;
                }
                
                totalWords++;
                wordLength = 0;
            }
        }
        
        if (wordLength > 0)
        {
            string word = Encoding.ASCII.GetString(wordBuffer.Slice(0, wordLength));
            if (counts.TryGetValue(word, out int count))
            {
                counts[word] = count + 1;
            }
            else
            {
                counts[word] = 1;
            }
            totalWords++;
        }
        
        return (counts, totalWords);
    }
    
    static void WriteOutputFile(string inputFile, List<KeyValuePair<string, int>> sorted, 
                                int totalWords, int uniqueWords, double executionTime)
    {
        string outputFile = Path.GetFileNameWithoutExtension(inputFile) + "_csharp_results.txt";
        
        using (var writer = new StreamWriter(outputFile))
        {
            writer.WriteLine("Word Frequency Analysis - C# Implementation");
            writer.WriteLine($"Input file: {inputFile}");
            writer.WriteLine($"Generated: {DateTime.Now:yyyy-MM-dd HH:mm:ss}");
            writer.WriteLine($"Execution time: {executionTime:F2} ms");
            writer.WriteLine();
            writer.WriteLine($"Total words: {totalWords:N0}");
            writer.WriteLine($"Unique words: {uniqueWords:N0}");
            writer.WriteLine();
            writer.WriteLine("Top 100 Most Frequent Words:");
            writer.WriteLine("Rank  Word            Count     Percentage");
            writer.WriteLine("----  --------------- --------- ----------");
            
            for (int i = 0; i < Math.Min(100, sorted.Count); i++)
            {
                var kvp = sorted[i];
                double percentage = (kvp.Value * 100.0) / totalWords;
                writer.WriteLine($"{i + 1,4}  {kvp.Key,-15} {kvp.Value,9:N0} {percentage,10:F2}%");
            }
        }
        
        Console.WriteLine($"\nResults written to: {outputFile}");
    }
}

File: wordcount.js
#!/usr/bin/env node
// wordcount.js - Word frequency counter
// Usage: node wordcount.js [filename]

const fs = require('fs');
const { performance } = require('perf_hooks');

const filename = process.argv[2] || 'book.txt';

if (!fs.existsSync(filename)) {
    console.error(`Error: File '${filename}' not found`);
    console.log('Usage: node wordcount.js [filename]');
    console.log('\nTo create a test file:');
    console.log('curl https://www.gutenberg.org/files/2701/2701-0.txt -o book.txt');
    process.exit(1);
}

console.log(`Processing file: ${filename}`);
const startTime = performance.now();
const startMemory = process.memoryUsage().heapUsed;

try {
    const stats = fs.statSync(filename);
    const buffer = Buffer.allocUnsafe(stats.size);
    const fd = fs.openSync(filename, 'r');
    fs.readSync(fd, buffer, 0, stats.size, 0);
    fs.closeSync(fd);
    
    const counts = new Map();
    let currentWord = '';
    let totalWords = 0;
    
    const isLetter = new Uint8Array(256);
    for (let i = 65; i <= 90; i++) isLetter[i] = 1;
    for (let i = 97; i <= 122; i++) isLetter[i] = 1;
    
    for (let i = 0; i < buffer.length; i++) {
        const char = buffer[i];
        
        if (isLetter[char]) {
            currentWord += String.fromCharCode(char >= 65 && char <= 90 ? char + 32 : char);
        } else if (currentWord.length > 0) {
            const count = counts.get(currentWord);
            counts.set(currentWord, count ? count + 1 : 1);
            totalWords++;
            currentWord = '';
        }
    }
    
    if (currentWord.length > 0) {
        const count = counts.get(currentWord);
        counts.set(currentWord, count ? count + 1 : 1);
        totalWords++;
    }
    
    // Use typed array for sorting to avoid object allocation
    const sorted = Array.from(counts.entries())
        .sort((a, b) => b[1] - a[1]);
    
    const endTime = performance.now();
    const endMemory = process.memoryUsage().heapUsed;
    const executionTime = (endTime - startTime).toFixed(2);
    const memoryUsed = ((endMemory - startMemory) / 1024 / 1024).toFixed(2);
    const fileSize = (stats.size / 1024 / 1024).toFixed(2);
    
    console.log('\n=== Top 10 Most Frequent Words ===');
    sorted.slice(0, 10).forEach(([word, count], index) => {
        console.log(`${(index + 1).toString().padStart(2)}. ${word.padEnd(15)} ${count.toLocaleString()}`);
    });
    
    console.log('\n=== Statistics ===');
    console.log(`File size:       ${fileSize} MB`);
    console.log(`Total words:     ${totalWords.toLocaleString()}`);
    console.log(`Unique words:    ${counts.size.toLocaleString()}`);
    console.log(`Execution time:  ${executionTime} ms`);
    console.log(`Memory used:     ${memoryUsed} MB`);
    console.log(`Node.js version: ${process.version}`);
    
    const outputFilename = filename.replace(/\.[^.]+$/, '') + '_javascript_results.txt';
    const outputContent = [
        'Word Frequency Analysis - JavaScript Implementation',
        `Input file: ${filename}`,
        `Generated: ${new Date().toISOString()}`,
        `Execution time: ${executionTime} ms`,
        '',
        `Total words: ${totalWords.toLocaleString()}`,
        `Unique words: ${counts.size.toLocaleString()}`,
        '',
        'Top 100 Most Frequent Words:',
        'Rank  Word            Count     Percentage',
        '----  --------------- --------- ----------',
        ...sorted.slice(0, 100).map(([word, count], index) => {
            const percentage = (count * 100 / totalWords).toFixed(2);
            return `${(index + 1).toString().padStart(4)}  ${word.padEnd(15)} ${count.toString().padStart(9)} ${percentage.padStart(10)}%`;
        })
    ].join('\n');
    
    fs.writeFileSync(outputFilename, outputContent);
    console.log(`\nResults written to: ${outputFilename}`);
    
} catch (error) {
    console.error('Error processing file:', error.message);
    process.exit(1);
}